### ISSUE TITLE
`get_mask_mod` and `get_score_mod` produces incorrect logical to physical translation.

### ISSUE THREAD

**Issue by man2machine** [meta-pytorch/attention-gym#189]
## TLDR

The `get_mask_mod` function in `paged_attention.py` incorrectly handles the `physical_to_logical` mapping when batch index slots are non-sequential, causing attention to read from wrong KV cache locations and producing garbage output for multi-request batches. The same error exists in the `get_score_mod` function.

## Long Description

When `convert_logical_block_mask` is called with `batch_idx` passed in, it slices the `page_table` to create a tensor aligned with the batch dimension. However, `get_mask_mod` does not slice `physical_to_logical`, causing a mismatch between the actual batch indices used for the prefill/decode batch and the batch indices used for the mask mod function.

Here is an example:

#### 2. Server calls decode with batch_idx = [3, 7, 1]

```
batch_idx = torch.tensor([3, 7, 1])
mask = self.get_decode_mask(batch_idx, input_pos)
```

#### 3. get_decode_mask slices the precomputed mask

```
new_kv_indices[:, :, :, :MAX_BLOCKS_IN_COL] = torch.gather(page_table, ...
```

Result for new mask:
- Position 0: data from batch slot 3
- Position 1: data from batch slot 7
- Position 2: data from batch slot 1

#### 4. convert_logical_block_mask slices page_table

```
page_table = self.page_table[batch_idx]
```

Result for sliced page table:
- Row 0: slot 3's page table
- Row 1: slot 7's page table
- Row 2: slot 1's page table

The conversion correctly maps logical to physical indices using this sliced page_table.

#### 5. get_mask_mod converts the mask mod to use physical instead of logic, but does not use the right batch indices.

```
def new_mask_mod(b, h, q_idx, physical_kv_idx):
    physical_kv_block = physical_kv_idx // self.page_size
    physical_kv_offset = physical_kv_idx % self.page_size
    logical_block_idx = self.physical_to_logical[b, physical_kv_block]  # Bug location
    ...
```

When FlexAttention executes:
- b=0 (position 0, should be slot 3): Reads `self.physical_to_logical[0, ...]` 
  - Should read slot 3's mapping, but reads slot 0's mapping instead
- b=1 (position 1, should be slot 7): Reads `self.physical_to_logical[1, ...]`
  - Should read slot 7's mapping, but reads slot 1's mapping instead
- b=2 (position 2, should be slot 1): Reads `self.physical_to_logical[2, ...]`
  - Should read slot 1's mapping, but reads slot 2's mapping instead

6. Result: Attention reads wrong KV cache locations

When actually running this using a real model, I found the output was garbage.

## How to fix

To fix you just need to simply slice `physical_to_logical` in the `get_mask_mod` and `get_score_mod` functions to align with the input batch indices. So you would do:

```
physical_to_logical_sliced = self.physical_to_logical[batch_idx]
# More lines of code...
logical_block_idx = physical_to_logical_sliced[b, physical_kv_block]
```

It took a lot of time to find this bug. Later on, I also found that flex-nano-vllm, which is a minimal inference server based on this repo already has this bug fixed. I'll put in a PR for this soon.

**Comment by drisspg** [comment]
fix has landed
